examples/policy_gradient/reinforce/agent.py

```diff
@@ -2,48 +2,185 @@
+                           "currently not supported")
+    self.prev_logits = tf.placeholder(tf.float32, shape=(None, self.logit_dim))
+
+    # Defines operations for staging input training data.
+    self.num_splits = len(devices)
+    with tf.device("/cpu:0"):
+      stage = StagingArea(
+          [self.observations.dtype, self.advantages.dtype,
+           self.actions.dtype, self.prev_logits.dtype],
+          [self.observations.shape, self.advantages.shape,
+           self.actions.shape, self.prev_logits.shape])
+    self.stage_ops = []
+    for obs, adv, acts, plgs in zip(
+        tf.split(self.observations, self.num_splits),
+        tf.split(self.advantages, self.num_splits),
+        tf.split(self.actions, self.num_splits),
+        tf.split(self.prev_logits, self.num_splits)):
+      p_op = stage.put([obs, adv, acts, plgs])
+      self.stage_ops.append(p_op)
+
+    # Defines the model replicas (i.e. "towers"), one per device.
+    self.ppo_towers = []
+    with tf.variable_scope("shared_policy_net"):
+      for i, device in enumerate(devices):
+        with tf.device(device):
+          obs, adv, acts, plgs = stage.get()
+          ppo = ProximalPolicyLoss(
+              self.env.observation_space, self.env.action_space, preprocessor,
+              obs, adv, acts, plgs, self.logit_dim, self.kl_coeff,
+              distribution_class, config, self.sess, report_metrics=i == 0)
+          self.ppo_towers.append(ppo)
+        tf.get_variable_scope().reuse_variables()
+    self.optimizer = tf.train.AdamOptimizer(config["sgd_stepsize"])
+    grads = []
+    for i, device in enumerate(devices):
+      with tf.name_scope("tower_" + str(i)):
+        with tf.device(device):
+          grads.append(self.optimizer.compute_gradients(self.ppo_towers[i].loss))
+
+    # The final training op which executes in parallel over the model towers.
+    average_grad = average_gradients(grads)
+    self.train_op = self.optimizer.apply_gradients(average_grad)
+
+    # Metric ops
+    with tf.name_scope("test_outputs"):
+      self.mean_loss = tf.reduce_mean(
+        tf.stack(values=[p.loss for p in self.ppo_towers]), 0)
+      self.mean_kl = tf.reduce_mean(
+        tf.stack(values=[p.mean_kl for p in self.ppo_towers]), 0)
+      self.mean_entropy = tf.reduce_mean(
+        tf.stack(values=[p.mean_entropy for p in self.ppo_towers]), 0)
+
+    # References to the model weights
+    self.variables = ray.experimental.TensorFlowVariables(
+        self.ppo_towers[0].loss,  # all towers have equivalent vars
+        self.sess)
+    self.observation_filter = MeanStdFilter(preprocessor.shape, clip=None)
+    self.reward_filter = MeanStdFilter((), clip=5.0)
     self.sess.run(tf.global_variables_initializer())
 
+  def stage_trajectory_data(self, batch):
+    self.sess.run(
+        self.stage_ops,
+        feed_dict={self.observations: make_divisible_by(batch["observations"], self.num_splits),
```

Comment:
this is potentially a little dangerous I think because it might drop data in every batch; I'd consider asserting that the batch size is a multiple of num_splits and either giving an error if it is not, or making it divisible and giving a warning